File: Sources/InstantTranslator/Services/TrialTranslationService.swift
import Foundation

enum TrialTranslationService {
    static let dailyLimit = 50
    static let maxCharacters = 3000

    private static let userDefaults = UserDefaults.standard

    static func remainingToday() -> Int {
        max(0, dailyLimit - usedToday())
    }

    static func canTranslate() -> Bool {
        usedToday() < dailyLimit
    }

    static func isTextWithinLimit(_ text: String) -> Bool {
        text.count <= maxCharacters
    }

    static func recordTranslation() {
        let key = todayKey()
        let current = userDefaults.integer(forKey: key)
        userDefaults.set(current + 1, forKey: key)
    }

    static func usedToday() -> Int {
        userDefaults.integer(forKey: todayKey())
    }

    private static func todayKey() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return "trialTranslations_\(formatter.string(from: Date()))"
    }
}

File: windows/ControlV.Core/Services/TrialTranslationService.cs
namespace ControlV.Core;

/// Local trial accounting (the server enforces the real limits; this avoids
/// pointless requests). Mirrors the Mac constants.
public sealed class TrialTranslationService
{
    public const int DailyLimit = 50;
    public const int MaxCharacters = 3000;

    private readonly IKeyValueStore _store;
    private readonly Func<DateTimeOffset> _now;

    public TrialTranslationService(IKeyValueStore store, Func<DateTimeOffset>? now = null)
    {
        _store = store;
        _now = now ?? (() => DateTimeOffset.Now);
    }

    public int UsedToday() => _store.GetInt(TodayKey());
    public int RemainingToday() => Math.Max(0, DailyLimit - UsedToday());
    public bool CanTranslate() => UsedToday() < DailyLimit;
    public static bool IsTextWithinLimit(string text) => text.Length <= MaxCharacters;
    public void RecordTranslation() => _store.SetInt(TodayKey(), UsedToday() + 1);

    private string TodayKey() => $"trialTranslations_{_now():yyyy-MM-dd}";
}
